### PR TITLE
Dmsj.c forest in thunder

### PR DESCRIPTION
**Pull request by dmsj**
Minor changes to test other planners in the `Thunder` framework

### CODE REVIEW

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp [hunk 1]
```diff
@@ -115,21 +115,24 @@ namespace ompl
             }
             else if (InformedSampler::space_->isCompound())
             {
+              std::cout << "PathLengthDirectInfSampler: Compound StateSpace. State Space Type: " << InformedSampler::space_->getType() << std::endl;
```
Comment (duanyutong) [inline src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:118]: there must be logging support in ompl i'm assuming?

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp [hunk 2]
```diff
@@ -115,21 +115,24 @@ namespace ompl
...
-                                        "2 subspaces.");
-                    }
+                    // // Sanity check
+                    // if (compoundSpace->getSubspaceCount() != 2u)
+                    // {
+                    //     std::cout << "state space has " << compoundSpace->getSubspaceCount() << " subspaces" << std::endl;
+                    //     // Pout
+                    //     throw Exception("The provided compound StateSpace is SE(2) or SE(3) but does not have exactly "
+                    //                     "2 subspaces.");
+                    // }
```
Comment (duanyutong) [inline src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:135]: why was the sanity check removed?

File: src/ompl/base/terminationconditions/src/CostConvergenceTerminationCondition.cpp [hunk 3]
```diff
@@ -57,11 +57,20 @@ void ompl::base::CostConvergenceTerminationCondition::processNewSolution(const o
     double costLowerThreshold = (1. - epsilon_) * averageCost_;
     double costUpperThreshold = (1. + epsilon_) * averageCost_;
     averageCost_ = newCost;
+    std::cout << "added a new solution with cost " << solutionCost.value() << " and average cost " << averageCost_
```
Comment (duanyutong) [inline src/ompl/base/terminationconditions/src/CostConvergenceTerminationCondition.cpp:60]: need to clean up the stdout after testing

File: src/ompl/tools/thunder/Thunder.h [hunk 4]
```diff
@@ -147,6 +147,15 @@ namespace ompl
                 return solve(ptc);
             }
 
+            base::PlannerStatus solve(const base::PlannerTerminationCondition &ptc, std::size_t minSolCount,
+                                      std::size_t maxSolCount,bool hybridize)
```
Comment (duanyutong) [inline src/ompl/tools/thunder/Thunder.h:151]: all const

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp (hunk 1 above)
Comment (jackievideira) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:118]: i don't think there's any that works, a lot of the ompl log lines i see are commented out... i'll remove all the std::out lines when we're ready to merge because they're really verbose and only useful for actively debugging ompl anyways

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp (hunk 2 above)
Comment (jackievideira) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp]: @dmsj i put this back in and didn't have any issues planning, is there any reason to keep it removed?

Comment (jackievideira) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp]: @yammineramy this is where you're throwing, i'll comment it back out

File: src/ompl/tools/thunder/Thunder.h [hunk 5]
```diff
@@ -88,13 +88,13 @@ namespace ompl
         public:
             /** \brief Constructor needs the state space used for planning. */
             explicit Thunder(const base::SpaceInformationPtr &si, double stretch_factor = 1.2, double DenseD = 0.001,
-                             double SparseD = 0.1);
+                             double SparseD = 0.04);
```
Comment (duanyutong) [inline src/ompl/tools/thunder/Thunder.h:91]: i think we should change this in our call, not in the default here

Comment (yammineramy) [reply src/ompl/tools/thunder/Thunder.h:91]: I can take care of that...

File: src/ompl/tools/multiplan/src/ParallelPlan.cpp [hunk 6]
```diff
@@ -172,8 +174,10 @@ void ompl::tools::ParallelPlan::solveMore(base::Planner *planner, std::size_t mi
         unsigned int nrSol = ++foundSolCount_;
         foundSolCountLock_.unlock();
 
-        if (nrSol >= maxSolCount)
-            ptc->terminate();
+        if (nrSol >= maxSolCount) {
```
Comment (aneksteind) [inline src/ompl/tools/multiplan/src/ParallelPlan.cpp:177]: let's undo this change, it just makes keeping up-to-date with the upstream more difficult

File: src/ompl/tools/thunder/Thunder.h [hunk 7]
```diff
@@ -245,6 +269,11 @@ namespace ompl
 
             bool hybridize_{true};
 
+            std::size_t minSolCount_ {21};
```
Comment (yammineramy) [inline src/ompl/tools/thunder/Thunder.h:272]: If we want a chance at hybridize = false to work as intended, minSolCount needs to be 1.
It is used for two things here, determine when we stop all threads and return if hybridize = false, and the number of solutions we start hybridizing after when hybridize = true.
If we care about the second use, we need to decouple the two uses.

File: src/ompl/base/src/StateSpace.cpp [hunk 8]
```diff
@@ -1071,8 +1071,8 @@ double ompl::base::CompoundStateSpace::distance(const State *state1, const State
     const auto *cstate2 = static_cast<const CompoundState *>(state2);
     double dist = 0.0;
     for (unsigned int i = 0; i < componentCount_; ++i)
-        dist += weights_[i] * components_[i]->distance(cstate1->components[i], cstate2->components[i]);
-    return dist;
+        dist += weights_[i] * std::pow(components_[i]->distance(cstate1->components[i], cstate2->components[i]), 2);
```
[the remark this answers is not in the document]
Comment (yammineramy) [reply src/ompl/base/src/StateSpace.cpp:1074]: We are only changing how we calculate distances in n dimensional spaces from L1 to L2. There is really no reason for it to be L1. What we can do is add an integer parameter that defaults to 1 and alter it, but I think leave it as is is not a problem.

File: src/ompl/tools/thunder/src/Thunder.cpp [hunk 9]
```diff
@@ -301,7 +319,7 @@ ompl::base::PlannerStatus ompl::tools::Thunder::solve(const base::PlannerTermina
                 log.is_saved = "less_2_states";
                 log.too_short = true;
             }
-            else if (false)  // always add when from recall
+            else if (true)  // always add when from recall
```
Comment (yammineramy) [inline src/ompl/tools/thunder/src/Thunder.cpp:322]: Should we make this into a flag?

Comment (jackievideira) [reply src/ompl/tools/thunder/src/Thunder.cpp:322]: Is there a performance difference/use for controlling this? I haven't experimented with the flag false vs true

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp [hunk 10]
```diff
@@ -117,19 +117,21 @@ namespace ompl
             {
                 // Check that it is SE2 or SE3
                 if (InformedSampler::space_->getType() == STATE_SPACE_SE2 ||
-                    InformedSampler::space_->getType() == STATE_SPACE_SE3)
+                    InformedSampler::space_->getType() == STATE_SPACE_SE3 || 
+                    InformedSampler::space_->getType() == 98)
```
[the remark this answers is not in the document]
Comment (dmsj) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:121]: We need some enum value

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp [hunk 11]
```diff
@@ -117,19 +117,21 @@ namespace ompl
...
-                    // Sanity check
+                    // Sanity check - Removed
+                    /*
                     if (compoundSpace->getSubspaceCount() != 2u)
                     {
                         // Pout
                         throw Exception("The provided compound StateSpace is SE(2) or SE(3) but does not have exactly "
                                         "2 subspaces.");
-                    }
+                    }*/
```
[the remark this answers is not in the document]
Comment (dmsj) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:134]: yes; this does not work with our compound space as we are combining SO2 with real vector spaces in a way not covered by this check.

File: src/ompl/tools/thunder/src/Thunder.cpp (hunk 9 above)
Comment (yammineramy) [reply src/ompl/tools/thunder/src/Thunder.cpp:322]: Yes, I can write a unit test for this. It needs to be true when we are building dat files most of the time. When running on robot this part of the code does not even run, so it does not matter. The only time we want it to be false is if we are building dat files quickly for some reason.

File: src/ompl/tools/thunder/Thunder.h (hunk 7 above)
Comment (jackievideira) [reply src/ompl/tools/thunder/Thunder.h]: I set it to one, we can do any further work in a subsequent PR :)

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (duanyutong) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:121]: yes please add a named constant at least for now in consistent style with upstream

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (duanyutong) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:134]: can we preserve the existing check and have it detect our unsupported new case and let it pass? we never want to remove any existing functionality from ompl because i wouldn't expect that to be accepted upstream.

File: src/ompl/base/src/StateSpace.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (duanyutong) [reply src/ompl/base/src/StateSpace.cpp:1074]: please add ample comments justifying the change. anything you change will be under scrutiny by upstream maintainers.

File: src/ompl/tools/thunder/Thunder.h [hunk 12]
```diff
@@ -147,6 +148,15 @@ namespace ompl
...
 
+            base::PlannerStatus solve(const base::PlannerTerminationCondition &ptc, const std::size_t minSolCount,
+                                      const std::size_t maxSolCount, const bool hybridize)
+            {
+                hybridize_ = hybridize;
+                minSolCount_ = minSolCount;
+                maxSolCount_ = maxSolCount;
+                return solve(ptc);
+            }
+
```
Comment (duanyutong) [inline src/ompl/tools/thunder/Thunder.h:159]: why is this overload necessary? docstring is missing

File: src/ompl/tools/thunder/Thunder.h [hunk 13]
```diff
@@ -161,6 +171,26 @@ namespace ompl
...
 
+            /** \brief set the from scratch planner to be CForest */
+            void setCforest() {
+              plan_with_cforest_ = true;
+            }
+
+            /** \brief set the from scratch planner to be RRT */
+            void setRRT() {
+              plan_with_cforest_ = false;
+            }
```
Comment (duanyutong) [inline src/ompl/tools/thunder/Thunder.h:182]: the implementations of these two functions aren't the same. why this asymmetric and counterintuitive pattern?

File: src/ompl/tools/thunder/Thunder.h [hunk 14]
```diff
@@ -161,6 +171,26 @@ namespace ompl
...
+
+            /** \brief Set the number of threads to use for planning. */
+            void setNumThreads(size_t n_threads) {
+              n_threads_ = n_threads;
+            }
```
Comment (duanyutong) [inline src/ompl/tools/thunder/Thunder.h:187]: input param should be const

File: src/ompl/tools/thunder/src/Thunder.cpp [hunk 15]
```diff
@@ -82,8 +84,20 @@ void ompl::tools::Thunder::setup()
...
+        }
+        // set the size of the planner vector
+        planner_vec_.clear();
+        if (plan_with_cforest_) {
+          std::vector<base::PlannerPtr> planner_vec {1};
+          planner_vec_ = planner_vec;
+        } else {
+          std::vector<base::PlannerPtr> planner_vec {n_threads_};
+          planner_vec_ = planner_vec;
+        }
```
Comment (duanyutong) [inline src/ompl/tools/thunder/src/Thunder.cpp:99]: how does this code work at all?

the original `planner_vec_` is a vec of pointers, but you are making a vector of a single int element, where the int is the number of threads.

File: src/ompl/tools/thunder/src/Thunder.cpp [hunk 16]
```diff
@@ -214,8 +232,8 @@ ompl::base::PlannerStatus ompl::tools::Thunder::solve(const base::PlannerTermina
     {
         OMPL_DEBUG("Thunder: stopping only after all threads report a solution");
     }
-
-    lastStatus_ = pp_->solve(ptc, hybridize_);  //, hybridize);
+    
+    lastStatus_ = pp_->solve(ptc, minSolCount_, maxSolCount_, hybridize_);  //, hybridize);
```
Comment (duanyutong) [inline src/ompl/tools/thunder/src/Thunder.cpp:236]: is the `hybridize_` setting documented anywhere?

File: src/ompl/tools/thunder/src/Thunder.cpp (hunk 15 above)
Comment (dmsj) [reply src/ompl/tools/thunder/src/Thunder.cpp:99]: https://stackoverflow.com/questions/11457571/how-to-set-initial-size-of-stdvector/11457629

it creates a vector filled with that number of (uninitialized) pointers

Comment (duanyutong) [reply src/ompl/tools/thunder/src/Thunder.cpp:99]: you have to use `()` not `{}`. i'm sure this is wrong.

Comment (dmsj) [reply src/ompl/tools/thunder/src/Thunder.cpp]: you are correct! need `(n_threads)` to correctly make more than 1 thread. I'm not sure why it didn't throw -- but I guess since it's a vector of pointer? I was assigning the address to a size_t??

Comment (duanyutong) [reply src/ompl/tools/thunder/src/Thunder.cpp]: > I'm not sure why it didn't throw

i bet the reason is that i wasn't working

Comment (yammineramy) [reply src/ompl/tools/thunder/src/Thunder.cpp]: @jackievideira  do you think this is why you were seeing a high number of solutions returning when hybridize was false?
Other things might be playing a role too (hybridize = false is not perfect and we return approximate solutions), but this could also be playing a role.

File: src/ompl/base/src/StateSpace.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (yammineramy) [reply src/ompl/base/src/StateSpace.cpp:1074]: Yes, this should be easy to justify... I will write an argument for it

File: src/ompl/tools/thunder/src/Thunder.cpp (hunk 16 above)
Comment (yammineramy) [reply src/ompl/tools/thunder/src/Thunder.cpp:236]: // There are two modes for running parallel plan - one in which both threads are run until they both return a result
    // and/or fail
    // The second mode stops with the first solution found - we want this one
    if (!hybridize_)
    {
        // If \e hybridize is false, when the first solution is found, the rest of the planners are stopped as well.
        OMPL_DEBUG("Thunder: stopping when first solution is found from either thread");
    }
    else
    {
        OMPL_DEBUG("Thunder: stopping only after all threads report a solution");
    }

File: src/ompl/tools/thunder/src/Thunder.cpp (hunk 9 above)
Comment (yammineramy) [reply src/ompl/tools/thunder/src/Thunder.cpp:322]: This can be kept as true for now, I have a flag for it in my PR, and I will worry about the unit test as soon as we are done with the more pressing things

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (jackievideira) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:121]: @dmsj - i am creating an enum for this but i don't know what to call it or what state space this is representing

[the remark this answers is not in the document]
Comment (jackievideira) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp:121]: `STATE_SPACE_COMBINED_SO2_REAL`?

File: src/ompl/base/src/StateSpace.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (yammineramy) [reply src/ompl/base/src/StateSpace.cpp:1076]: unit tests will have to follow later

File: src/ompl/tools/thunder/src/Thunder.cpp (hunk 16 above)
Comment (jackievideira) [reply src/ompl/tools/thunder/src/Thunder.cpp:236]: in ParallelPlan.h it's documented

File: src/ompl/tools/thunder/Thunder.h (hunk 13 above)
Comment (jackievideira) [reply src/ompl/tools/thunder/Thunder.h:183]: what would be better here - a bool for each type? or should i create an enum value here? we use this to determine which planner type to make in setup() so i can't store the name of the planner or just create a planner type on these set functions, we also use it to determine the size of the planner ptr vec

File: src/ompl/tools/thunder/Thunder.h (hunk 12 above)
Comment (jackievideira) [reply src/ompl/tools/thunder/Thunder.h:160]: i added a docstring but i'm not sure it answers your question

File: src/ompl/tools/thunder/src/Thunder.cpp (hunk 9 above)
Comment (jackievideira) [reply src/ompl/tools/thunder/src/Thunder.cpp:323]: added a todo for you, post this PR

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp (hunk 2 above)
Comment (jackievideira) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp]: i put an if statement around this if we're in the new state space to skip it

File: src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (jackievideira) [reply src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp]: i added an if statement to check the state space we're in and skip if we're in the new one

File: src/ompl/tools/thunder/Thunder.h (hunk 13 above)
Comment (duanyutong) [reply src/ompl/tools/thunder/Thunder.h:183]: i'd suggest enums because there are two (maybe more in the future) named choices that are mutually exclusive as suggested by the setter names

if they aren't mutually exclusive and cforest is merely an addon option for RRT that's always used, then we keep the bool and rename the setters.

File: .vscode/settings.json [hunk 17]
```diff
@@ -0,0 +1,61 @@
+{
```
Comment (brook-stevens) [inline .vscode/settings.json:1]: I think this should be removed....

File: src/ompl/tools/thunder/Thunder.h [hunk 18]
```diff
@@ -255,6 +255,14 @@ namespace ompl
                 return DenseD_;
             };
 
+            void setSavePlansFromRecall (bool savePlansFromRecall) {
```
Comment (duanyutong) [inline src/ompl/tools/thunder/Thunder.h:258]: input arg should be const

Comment (yammineramy) [reply src/ompl/tools/thunder/Thunder.h]: fixed

File: src/ompl/tools/thunder/src/Thunder.cpp (hunk 15 above)
Comment (duanyutong) [reply src/ompl/tools/thunder/src/Thunder.cpp]: more discussion on slack 
https://dexai.slack.com/archives/CPZPT2S2W/p1642537366146500?thread_ts=1642089257.062900&cid=CPZPT2S2W

File: src/ompl/tools/thunder/Thunder.h (hunk 13 above)
Comment (jackievideira) [reply src/ompl/tools/thunder/Thunder.h]: ok - added an enum thunderPlanner, was trying to keep style similar to what's in OMPL already and half copied this from OptimalPlanning.cpp which names an optimalPlanner enum type

File: src/ompl/base/ProblemDefinition.h [hunk 19]
```diff
@@ -404,6 +404,16 @@ namespace ompl
             /** \brief Print information about the start and goal states and the optimization objective */
             void print(std::ostream &out = std::cout) const;
 
+            /** \brief Returns the flag that controls whether or not we accept approximate solutions */
+            bool acceptsApproximate() {
+                return acceptApproximate_;
+            }
+
+            /** \brief Sets the flag that controls whether or not we accept approximate solutions */
+            void setAcceptApproximate(bool acceptApproximate) {
```
Comment (duanyutong) [inline src/ompl/base/ProblemDefinition.h:413]: const bool acceptApproximate

Comment (duanyutong) [inline src/ompl/base/ProblemDefinition.h:413]: const bool acceptApproximate

**Review by duanyutong — COMMENTED** [review]
lgtm

**Review by duanyutong — APPROVED** [review]
